docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+# CIP: Pipelines, The Registry and Storage
+
+## Status
+
+Current Status: `Under Discussion`
+
+## Motivation
+
+The top-line motivation behind this CIP is to provide a generic system for preprocessing data before
+it is indexed by Chroma. This system should be flexible enough to allow users to
+define their own preprocessing pipelines, while also providing a set of default
+pipelines that can be used out of the box. The abstractions proposed in this CIP are born out of observing
+several problems the Chroma community has encountered while using the current system.
+
+1. Embedding Function Persistence
+
+Today, Chroma allows users to implicitly transform the data they provide to add() by embedding it. For example,
+users can call something akin to:
+
+```python
+collection = client.create_collection("my_collection", embedding_function=OpenAIEmbeddingFunction())
+collection.add(ids, documents=["text_to_embed"])
+```
+
+One key problem with this approach is that this association between the embedding_function
+and the collection is not persisted. So users are forced to remember which embedding_function
+they used when they want to query the collection. This is not ideal. Concretely, this means
+users do something like this:
+
+```python
+collection = client.create_collection("my_collection", embedding_function=OpenAIEmbeddingFunction())
+collection.add(ids, documents=["text_to_embed"])
+
+# ... some time later
+
+collection = client.get_collection("my_collection", embedding_function=OpenAIEmbeddingFunction())
+collection.query(documents=["text_to_query"])
+```
+
+This mechanism is error-prone, confusing and verbose. We would prefer a system that
+allows for persisting the association between the embedding_function and the collection.
+
+2. Chunking
+
+When using Chroma, most users first chunk their documents into sizes that are appropriate
+for the respective embedding model they are using. This chunking strategy is handled outside of
+Chroma, and is often done in a way that is specific to the embedding model. Oftentimes users want to
+maintain the association between the chunks and the original document. Today, users accomplish this by manually
+chunking their documents outside of Chroma and then passing the chunks to Chroma, with the association
+between the chunks and the original document being maintained in the metadata of the document
+or by the user themselves. Users often forget to maintain this association, find it cumbersome
+to do so, or want to iterate on this in a versioned way.
+
+3. Loading & Storing documents
+
+Users often want to load documents from a variety of sources. For example, users may want to load
+documents from a database, a file system, or a remote service, these documents may be in a variety
+of formats, such as text, images, or audio. Today, users are responsible for loading these documents
```

Comment:
> responsible for loading
and keeping them up-to-date. e.g. triggering a sync, which brings some more challenges.

Reply:
also this!